File: sdks/dotnet/Keepty/Protocol.cs
using System;
using System.Buffers.Binary;
using System.IO;
using System.Net.Sockets;

namespace Keepty;

/// <summary>Client role when connecting to a keepty broker.</summary>
public enum Role : byte
{
    /// <summary>Exclusive: can send input + resize.</summary>
    Writer = 1,
    /// <summary>Read-only: receives output broadcast.</summary>
    Watcher = 2,
    /// <summary>Read-only: for programmatic screen analysis.</summary>
    Monitor = 3,
}

/// <summary>Message types in the keepty protocol.</summary>
public enum MsgKind : byte
{
    Hello = 1,
    HelloAck = 2,
    Input = 3,
    Output = 4,
    Resize = 5,
    ResizeAck = 6,
    Exit = 10,
    Shutdown = 11,
    Ping = 12,
    Pong = 13,
    Error = 127,
}

/// <summary>A keepty protocol frame.</summary>
public record Frame(MsgKind Kind, byte[] Payload)
{
    /// <summary>Deprecated: use SocketDirResolved() instead.</summary>
    public static readonly string SocketDir = "/tmp";

    /// <summary>Resolve the runtime socket directory.</summary>
    public static string SocketDirResolved()
    {
        var xdg = Environment.GetEnvironmentVariable("XDG_RUNTIME_DIR");
        if (!string.IsNullOrEmpty(xdg))
            return Path.Combine(xdg, "keepty");
        return Path.Combine(Path.GetTempPath().TrimEnd(Path.DirectorySeparatorChar), "keepty");
    }
    public const int MaxPayloadSize = 1_048_576; // 1MB
    public const int MaxFrameBodySize = MaxPayloadSize + 2; // payload + version + kind

    /// <summary>Construct the broker socket path for a session.</summary>
    public static string SocketPath(string sessionId) =>
        Path.Combine(SocketDirResolved(), $"keepty-{sessionId}.sock");

    /// <summary>Encode to wire format: [u32 len BE][u8 version][u8 kind][payload]</summary>
    public byte[] Encode()
    {
        int frameLen = 2 + Payload.Length;
        byte[] buf = new byte[4 + frameLen];
        BinaryPrimitives.WriteUInt32BigEndian(buf.AsSpan(0, 4), (uint)frameLen);
        buf[4] = 1; // protocol version
        buf[5] = (byte)Kind;
        Payload.CopyTo(buf.AsSpan(6));
        return buf;
    }

    /// <summary>Read one frame from a stream. Returns null on EOF.</summary>
    public static Frame? ReadFrom(Stream stream)
    {
        byte[] lenBuf = new byte[4];
        if (!ReadExact(stream, lenBuf)) return null;

        uint frameLen = BinaryPrimitives.ReadUInt32BigEndian(lenBuf);
        if (frameLen < 2)
            throw new InvalidDataException($"Frame too short: {frameLen}");
        if (frameLen > (uint)MaxFrameBodySize)
            throw new InvalidDataException($"Frame too large: {frameLen} bytes (max {MaxFrameBodySize})");

        byte[] data = new byte[frameLen];
        if (!ReadExact(stream, data))
            throw new EndOfStreamException("Connection closed mid-frame");

        byte version = data[0];
        if (version != 1)
            throw new InvalidDataException($"Unsupported protocol version: {version} (expected 1)");
        var kind = (MsgKind)data[1];
        byte[] payload = new byte[frameLen - 2];
        Array.Copy(data, 2, payload, 0, payload.Length);
        return new Frame(kind, payload);
    }

    /// <summary>Write this frame to a stream.</summary>
    public void WriteTo(Stream stream)
    {
        byte[] encoded = Encode();
        stream.Write(encoded);
        stream.Flush();
    }

    private static bool ReadExact(Stream stream, byte[] buffer)
    {
        int offset = 0;
        while (offset < buffer.Length)
        {
            int read = stream.Read(buffer, offset, buffer.Length - offset);
            if (read == 0)
            {
                if (offset == 0) return false;
                throw new EndOfStreamException($"Connection closed (got {offset}/{buffer.Length} bytes)");
            }
            offset += read;
        }
        return true;
    }
}

/// <summary>Message encoding/decoding helpers.</summary>
public static class Messages
{
    public static byte[] EncodeHello(Role role, ushort cols, ushort rows)
    {
        byte[] buf = new byte[5];
        buf[0] = (byte)role;
        BinaryPrimitives.WriteUInt16BigEndian(buf.AsSpan(1, 2), cols);
        BinaryPrimitives.WriteUInt16BigEndian(buf.AsSpan(3, 2), rows);
        return buf;
    }

    public static (uint pid, ushort cols, ushort rows) DecodeHelloAck(byte[] payload) => (
        BinaryPrimitives.ReadUInt32BigEndian(payload.AsSpan(0, 4)),
        BinaryPrimitives.ReadUInt16BigEndian(payload.AsSpan(4, 2)),
        BinaryPrimitives.ReadUInt16BigEndian(payload.AsSpan(6, 2))
    );

    public static byte[] EncodeResize(ushort cols, ushort rows)
    {
        byte[] buf = new byte[4];
        BinaryPrimitives.WriteUInt16BigEndian(buf.AsSpan(0, 2), cols);
        BinaryPrimitives.WriteUInt16BigEndian(buf.AsSpan(2, 2), rows);
        return buf;
    }

    public static (Role role, ushort cols, ushort rows) DecodeHello(byte[] payload) => (
        (Role)payload[0],
        BinaryPrimitives.ReadUInt16BigEndian(payload.AsSpan(1, 2)),
        BinaryPrimitives.ReadUInt16BigEndian(payload.AsSpan(3, 2))
    );

    public static byte[] EncodeResizeAck(uint gen, ushort cols, ushort rows)
    {
        byte[] buf = new byte[8];
        BinaryPrimitives.WriteUInt32BigEndian(buf.AsSpan(0, 4), gen);
        BinaryPrimitives.WriteUInt16BigEndian(buf.AsSpan(4, 2), cols);
        BinaryPrimitives.WriteUInt16BigEndian(buf.AsSpan(6, 2), rows);
        return buf;
    }

    public static (uint gen, ushort cols, ushort rows) DecodeResizeAck(byte[] payload) => (
        BinaryPrimitives.ReadUInt32BigEndian(payload.AsSpan(0, 4)),
        BinaryPrimitives.ReadUInt16BigEndian(payload.AsSpan(4, 2)),
        BinaryPrimitives.ReadUInt16BigEndian(payload.AsSpan(6, 2))
    );

    public static int DecodeExit(byte[] payload) =>
        BinaryPrimitives.ReadInt32BigEndian(payload.AsSpan(0, 4));
}

File: sdks/python/keepty/protocol.py
"""
keepty wire protocol — pure Python implementation.

Length-prefixed binary frames over Unix sockets.
Frame: [u32 len BE][u8 version][u8 kind][payload...]
"""

import os
import struct
from enum import IntEnum
from typing import Optional


PROTOCOL_VERSION = 1
SOCKET_DIR = "/tmp"  # Deprecated: use socket_dir()
MAX_PAYLOAD_SIZE = 1_048_576  # 1MB
MAX_FRAME_BODY_SIZE = MAX_PAYLOAD_SIZE + 2  # payload + version + kind


def socket_dir() -> str:
    """Resolve the runtime socket directory.
    Priority: $XDG_RUNTIME_DIR/keepty → $TMPDIR/keepty (via tempfile)
    """
    xdg = os.environ.get("XDG_RUNTIME_DIR", "")
    if xdg:
        return os.path.join(xdg, "keepty")
    import tempfile
    return os.path.join(tempfile.gettempdir(), "keepty")


class Role(IntEnum):
    """Client role when connecting to a keepty broker."""
    WRITER = 1   # Exclusive: can send input + resize
    WATCHER = 2  # Read-only: receives output broadcast
    MONITOR = 3  # Read-only: for programmatic screen analysis


class MsgKind(IntEnum):
    """Message types in the keepty protocol."""
    HELLO = 1
    HELLO_ACK = 2
    INPUT = 3
    OUTPUT = 4
    RESIZE = 5
    RESIZE_ACK = 6
    EXIT = 10
    SHUTDOWN = 11
    PING = 12
    PONG = 13
    ERROR = 127


class Frame:
    """A keepty protocol frame."""

    __slots__ = ("kind", "payload")

    def __init__(self, kind: MsgKind, payload: bytes = b""):
        self.kind = kind
        self.payload = payload

    def encode(self) -> bytes:
        """Encode to wire format: [u32 len BE][u8 version][u8 kind][payload]"""
        frame_len = 2 + len(self.payload)  # version + kind + payload
        return struct.pack(">IB", frame_len, PROTOCOL_VERSION) + bytes([self.kind]) + self.payload

    @staticmethod
    def read_from(sock) -> Optional["Frame"]:
        """Read one frame from a socket. Returns None on EOF."""
        len_buf = _recv_exact(sock, 4)
        if len_buf is None:
            return None
        frame_len = struct.unpack(">I", len_buf)[0]
        if frame_len < 2:
            raise ValueError(f"Frame too short: {frame_len}")
        if frame_len > MAX_FRAME_BODY_SIZE:
            raise ValueError(f"Frame too large: {frame_len} bytes (max {MAX_FRAME_BODY_SIZE})")
        data = _recv_exact(sock, frame_len)
        if data is None:
            return None
        version = data[0]
        if version != PROTOCOL_VERSION:
            raise ValueError(f"Unsupported protocol version: {version} (expected {PROTOCOL_VERSION})")
        kind = MsgKind(data[1])
        payload = data[2:]
        return Frame(kind, payload)

    def write_to(self, sock) -> None:
        """Write this frame to a socket."""
        sock.sendall(self.encode())

    def __repr__(self) -> str:
        return f"Frame({self.kind.name}, {len(self.payload)} bytes)"


def socket_path(session_id: str) -> str:
    """Construct the broker socket path for a session."""
    return os.path.join(socket_dir(), f"keepty-{session_id}.sock")


# --- Message helpers ---

def encode_hello(role: Role, cols: int, rows: int) -> bytes:
    return struct.pack(">BHH", role, cols, rows)


def decode_hello(payload: bytes) -> tuple[Role, int, int]:
    role_val, cols, rows = struct.unpack(">BHH", payload[:5])
    return Role(role_val), cols, rows


def encode_hello_ack(pty_pid: int, cols: int, rows: int) -> bytes:
    return struct.pack(">IHH", pty_pid, cols, rows)


def decode_hello_ack(payload: bytes) -> tuple[int, int, int]:
    pid, cols, rows = struct.unpack(">IHH", payload[:8])
    return pid, cols, rows


def encode_resize(cols: int, rows: int) -> bytes:
    return struct.pack(">HH", cols, rows)


def decode_resize(payload: bytes) -> tuple[int, int]:
    cols, rows = struct.unpack(">HH", payload[:4])
    return cols, rows


def encode_exit(code: int) -> bytes:
    return struct.pack(">i", code)


def decode_exit(payload: bytes) -> int:
    return struct.unpack(">i", payload[:4])[0]


def _recv_exact(sock, n: int) -> Optional[bytes]:
    """Read exactly n bytes from a socket. Returns None on EOF."""
    buf = bytearray()
    while len(buf) < n:
        chunk = sock.recv(n - len(buf))
        if not chunk:
            if not buf:
                return None
            raise ConnectionError(f"Connection closed mid-read (got {len(buf)}/{n} bytes)")
        buf.extend(chunk)
    return bytes(buf)
